File: lib/merge_sort.rb
require 'colorize'
require 'benchmark'

puts "Merge Sort Algorithm\n".bold

def sort_two(left, right)
  array = []
  l = 0
  r = 0
  loop do
    # Exit loop if end of both arrays has been reached
    break if l >= left.length && r >= right.length
    # If right array is finished OR (left array isnt finished AND left is less than right)
    if r >= right.length || (l < left.length && left[l] < right[r])
      array << left[l]
      l += 1
    else
      array << right[r]
      r += 1
    end
  end
  array
end

def recursive_split(array)
  # Break out of recursion if array length is 1
  return array if array.length <= 1
  # Slice array in half
  halves = array.each_slice((array.length/2.0).round).to_a
  # Recurse with new array
  left = recursive_split(halves[0])
  right = recursive_split(halves[1])
  # Perform sort action on results
  sort_two(left, right)
end

puts 'Uses recursion to split the array in half until left with only one item'
puts 'For each leaf node, the numbers are sorted by comparing one at at time between each array'
puts 'The smaller value is shovelled into the returned results, the index incremented and then checked again'
puts "Each returned array is then recompared with the next \"batch\"\n"

puts Benchmark.measure {
  array = []
  10000.times{ array << rand(10000) }
  recursive_split(array)
}

File: c/mergesort.c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>

/* When recursion reaches maximum, return pointer to array and size in memory
 * Do first for left, then right, concatenate arrays into new one
 * Wipe memory of left and right, return new pointer to concatenated array */

/* Things to work out:
 * 1. Split array in half correctly
 * 2. Return two values from a function? */

int get_input() {
  printf("Input limit:\n");
  int input;
  scanf("%d", &input);
  return input;
}

int recurse(int array, int size) {
  if (size <= 1) {
    return array;
  }
  int split_l = size/2;
  int left[split_l];
  for (int i = 0; i < size/2; i++) {
    left[i] = array[i];
  }
  recurse(left, split_l);
  return 0;
}


int main() {
  int input = get_input();
  int array[input];
  for (int i = 0; i < input; i++ ) {
    array[i] = (rand() % input);
  };
  int size = sizeof(array) / sizeof(int);
  recurse(array, size);
  return 0;
}


